apps/browser/bookmarks/bookmarkfiltermodel.cpp

```diff
@@ -28,24 +28,53 @@ bool BookmarkFilterModel::filterAcceptsRow(int sourceRow, const QModelIndex &sou
 {
     QModelIndex index = sourceModel()->index(sourceRow, 0, sourceParent);
 
-    if (sourceModel()->data(index, DeclarativeBookmarkModel::UrlRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)
-            || sourceModel()->data(index, DeclarativeBookmarkModel::TitleRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)) {
+    if ((sourceModel()->data(index, DeclarativeBookmarkModel::UrlRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)
+         || sourceModel()->data(index, DeclarativeBookmarkModel::TitleRole).toString().trimmed().contains(search(), Qt::CaseInsensitive))
+         && m_countFilterAccepts < m_maxDisplayedItems) {
+        m_countFilterAccepts++;
         return true;
     }
     return false;
 }
 
+void BookmarkFilterModel::setSourceModel(QAbstractItemModel *sourceModel)
+{
+    if (sourceModel) {
+        m_countFilterAccepts = 0;
+        m_maxDisplayedItems = sourceModel->rowCount();
+        QSortFilterProxyModel::setSourceModel(sourceModel);
+        invalidate();
+    }
+}
+
 QString BookmarkFilterModel::search() const
 {
     return m_search;
 }
 
+int BookmarkFilterModel::maxDisplayedItems() const
+{
+    return m_maxDisplayedItems;
+}
+
 void BookmarkFilterModel::setSearch(const QString &search)
 {
     if (m_search == search)
         return;
 
+    m_countFilterAccepts = 0;
     m_search = search;
+    invalidate();
```

Comment:
Rather than invalidating the entire model each time the search term changes, it would be better to perform a diff and only update the rows that have been added/removed. For example see the code in [`DeclarativeHistoryModel::updateModel()`](https://github.com/sailfishos/sailfish-browser/blob/1cd9c437f411a9159c7f1551f339522a61a6e2d0/apps/history/declarativehistorymodel.cpp#L113).

Reply:
@llewelld  `invalidate()` it was before the changes https://github.com/sailfishos/sailfish-browser/blob/1cd9c437f411a9159c7f1551f339522a61a6e2d0/apps/browser/bookmarks/bookmarkfiltermodel.cpp#L50

Reply:
Apologies for not explaining clearly. As I understand it, using `QAbstractListModel` would allow the model to be manipulated by Adding/Removing rows rather than invalidating the entire model each time the search term changes. This is the approach that's used for the [history model](https://github.com/sailfishos/sailfish-browser/blob/1cd9c437f411a9159c7f1551f339522a61a6e2d0/apps/history/declarativehistorymodel.cpp) for example. In that case, when the search term changes, a diff is performed in the `updateModel()` method and only changes are emitted.

It would be nice to do the same with the bookmarks, but it's not essential for your changes here.

Reply:
@llewelld you mind filing a task about it. Would be good to get it fixed for the next release. Should not be very big task to implement.

Reply:
@llewelld  `invalidateFilter ()` aren't these operations optimized? Why create another `QAbstractListModel` that will hold the filtered items.  `QSortFilterProxyMode` for this it is also intended.

Reply:
Yes, you're right @aroditelev and I was mistaken. If you use `invalidateFilter()` it'll be optimised, so in that case there's only limited benefit to changing to `QAbstractListModel()` (`QSortFilterProxyModel` doesn't know when it's a narrowing of the search term, but I guess the benefit that can be gained from that is small).

Reply:
Done